# Dependencies

In [ ]:
!!pip install geopy

['Requirement already satisfied: geopy in /usr/local/lib/python3.10/dist-packages (2.4.1)',
 'Requirement already satisfied: geographiclib<3,>=1.52 in /usr/local/lib/python3.10/dist-packages (from geopy) (2.0)']

In [ ]:
!pip install openml

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.0/158.0 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.9/93.9 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 15.3 MB/s eta 0:00:00
  Created wheel for liac-arff: filename=liac_arff-2.5.0-py3-none-any.whl size=11716 sha256=aa2810713930b75ebfe579b8382244fef981a4d7caa32023db84c4f491fb4dbc
  Stored in directory: /root/.cache/pip/wheels/5d/2a/9c/3895d9617f8f49a0883ba686326d598e78a1c2f54fe3cae86d
Successfully built liac-arff


In [ ]:
!pip install "vegafusion[embed]"

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.1/25.1 MB 15.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 658.1/658.1 kB 29.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.1/30.1 MB 28.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 206.3/206.3 kB 13.1 MB/s eta 0:00:00
  Attempting uninstall: altair
    Found existing installation: altair 4.2.2
    Uninstalling altair-4.2.2:
      Successfully uninstalled altair-4.2.2


In [ ]:
!pip install scikit_posthocs

In [ ]:
import pandas as pd
import geopandas as gpd
import numpy as np
import geopy
import altair as alt
import openml
from scipy.stats import kruskal
from scipy.stats import mannwhitneyu
import scikit_posthocs as sp

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
import plotly.graph_objects as go

from tqdm import tqdm
import time

alt.data_transformers.enable("vegafusion")
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('vegafusion')

In [ ]:
dataset = openml.datasets.get_dataset(42080)
df, _, _, _ = dataset.get_data(dataset_format='dataframe')

In [ ]:
!wget https://www2.census.gov/geo/tiger/GENZ2018/shp/cb_2018_us_state_500k.zip
!unzip cb_2018_us_state_500k.zip

state_boundaries = gpd.read_file("/content/cb_2018_us_state_500k.shp", columns=['STUSPS', 'NAME', 'geometry'])
state_boundaries.rename(columns={'STUSPS': 'state',
                           'Name': 'state_name'}, inplace=True)

df_EUA = state_boundaries.merge(df, on='state', how='inner')

--2024-11-12 15:29:45--  https://www2.census.gov/geo/tiger/GENZ2018/shp/cb_2018_us_state_500k.zip
Resolving www2.census.gov (www2.census.gov)... 23.198.156.203, 2600:1409:12:3a2::208c, 2600:1409:12:392::208c
Connecting to www2.census.gov (www2.census.gov)|23.198.156.203|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘cb_2018_us_state_500k.zip’

cb_2018_us_state_50     [       <=>          ]   3.15M  2.18MB/s    in 1.4s    

2024-11-12 15:29:47 (2.18 MB/s) - ‘cb_2018_us_state_500k.zip’ saved [3304931]

Archive:  cb_2018_us_state_500k.zip
  inflating: cb_2018_us_state_500k.shp.ea.iso.xml  
  inflating: cb_2018_us_state_500k.shp.iso.xml  
  inflating: cb_2018_us_state_500k.shp  
  inflating: cb_2018_us_state_500k.shx  
  inflating: cb_2018_us_state_500k.dbf  
  inflating: cb_2018_us_state_500k.prj  
 extracting: cb_2018_us_state_500k.cpg  


# Data Profilling


## **An Exploratory Data Analysis of U.S Individual Donations for Political Campaigs**


---
A Comissão Federal de Eleições (FEC) é uma agência independente que visa mediar e regular o financiamente de campanhas presidencias dos EUA. Sob o pretexto de registrar cada doação dos cidadãos, criou-se a base que será aqui trabalhada. A partir das bibliotecas **Vega-Altair** e **Plotly**, explora-se a criação de visualizações para compreender e investigar o comportamento das contribuições políticas ao redor do país.


---



### **As variáveis da base**

#### Variáveis Preditivas

* **cmte_id**: ID do comite para o qual a contribuição foi feita. Cada comitê possui um identificador único. Neste dataset estão presentes 7112 comites diferentes.

* **amndt_ind**: Indicador de Emenda. Tal variável é usada para mostrar se uma transação ou um registro é uma emenda a um registro anteriormente apresentado, sendo classificada por uma das três opções:
  * N - Novo registro;

  * A - Emenda a um registro anterior;
  * C - Correção/alteração num registro.

* **rpt_tp**: Tipo de Relatório de contribuição do Comitê ao qual o contribuinte está associado. Seus tipos são comumente diferenciados pela sua recorrência(mensal, trimestral, pré-eleição e etc).

* **transaction_pgi**: Indicador referente a qual etapa do ciclo eleitoral a contribuição em questão está associada.
  *  G - General Election, fase final do ciclo, onde um candidato é eleito para o cargo
  *  P - Primary Election, refere-se às eleições que são internas aos partidos
  *  S - Special Election, eleições convocadas fora do ciclo eleitoral regular para preencher vagas inesperadas
  *  E - Election,
  *  C - Convenction, Convenções Partidárias, que podem ser usadas para selecionar candidatos ou tomar outras decisões internas do partido.
  * O - Other,

  * R - Runoff Election (Segundo Turno)

  * 0 - ??

* **image_num**: Identificador único associado à imagem digitalizada do relatório financeiro.

* **transaction_tp**
  * 10 - Contribuição Financeira direta de um indivíduo ou entidade para um comitê de campanha de um candidato
  * 11 - Contribuição de uma Tribo Nativa
  * 15 - Contribuição financeira direta feita por um indivíduo, sociedade ou LLC (Empresa de responsabilidade limitada) para comitês políticos tradicionais, excluindo Super PACs e Hybrid PACs.
  * 15C - Contribuição de candidatos
  * 15E - Contribuições Direcionadas feitas por indivíduos, sociedades ou LLC para comitês políticos tradicionais
  * 19 - Doação para comunicação eleitoral

  * 20Y - Reembolso de Fundos não Eleitorais. Devolução de fundos que foram originalmente atribuídos a atividades não eleitorais.
  * 21Y - Reembolso de uma Tribo Nativa
  * 22Y - Reembolso de uma contribuição de uma indivíduo, parceiros ou uma empresa de responsabilidade limitada
  * 24I - Contribuição Direcionada por cheque
  * 24T - Contribuição Direcionada utilizando fundos do tesouro do comitê intermediário

* **entity_tp**
  * IND - Individual
  * ORG - Organização
  * CCM - Comitê do Candidato
  * PAC - Comitê de Ação Política
  * CAN - Candidato
  * COM - Comitê

  * PTY - Comitê Oficial do Partido

* **name** - Nome do Contribuinte
* **city** - Cidade do Contribuinte

* **state** - Estado do Contribuinte

* **zip_code** - Código Postal do Contribuinte

* **employer** - Empregador do Contribuinte

* **ocuppation** - Emprego do Contribuinte

* **transaction_dt** - Data de transação das contribuições

* **other_id** - ID de identificação de uma terceira parte envolvida na transação

* **tran_id** - ID único da transação

* **file_num** - Número do arquivo

* **memo_cd** - Identificador associado a presença de um memorando na transação, o qual fornece um contexto adicional à transação

* **memo_text** - Texto do memorando

*  **sub_id** - Identificador de submissão

#### Variável Alvo
* **transaction_amt**: O valor da transação em dólares. Indica a quantidade de dinheiro envolvida em uma contrbuição individual para campanhas políticas.

Assim, o dataset é composto por 5 variáveis de Identificação única, 13 variáveis categóricas, uma variável de texto, uma variável de data e uma varável numérica (alvo)

In [ ]:
df.shape

(3348209, 21)

In [ ]:
# Número de categorias únicas das variáveis categóricas
categorical_cols = df.select_dtypes(include='object')
num_categories = categorical_cols.nunique().sort_values(ascending=False)
print(num_categories)

tran_id            2999272
name               1534998
employer            657449
occupation          145304
zip_code             86334
city                 27709
cmte_id               7112
other_id              2543
state                   67
rpt_tp                  26
transaction_tp          11
transaction_pgi          8
entity_tp                7
amndt_ind                3
memo_cd                  2
dtype: int64


# Pre-Processing

## Handling Missing Values

In [ ]:
missing_percent = df.isna().sum().sort_values(ascending=False) / df.shape[0] * 100
missing_percent = missing_percent[missing_percent > 0]
print(missing_percent)

memo_cd            97.314325
other_id           92.992313
memo_text          87.438747
transaction_pgi    28.104966
employer           10.371216
occupation          5.546966
state               0.258556
city                0.064990
entity_tp           0.042948
transaction_dt      0.010573
tran_id             0.009766
name                0.004211
file_num            0.000030
dtype: float64


In [ ]:
#@title Co-ocorrência das variáveis com valores faltantes significantes
significant_missing_columns = ['occupation', 'employer', 'transaction_pgi', 'memo_cd', 'memo_text', 'other_id']
co_occur = df[significant_missing_columns].isna().astype(int)
co_occur_agg = co_occur.groupby(significant_missing_columns).size().reset_index(name='count').sort_values(by='count', ascending=False)
co_occur_agg.head(15)

occupation  employer  transaction_pgi  memo_cd  memo_text  other_id  \
7            0         0                0        1          1         1   
15           0         0                1        1          1         1   
54           1         1                0        1          1         1   
5            0         0                0        1          0         1   
23           0         1                0        1          1         1   
4            0         0                0        1          0         0   
12           0         0                1        1          0         0   
13           0         0                1        1          0         1   
31           0         1                1        1          1         1   
1            0         0                0        0          0         1   
62           1         1                1        1          1         1   
6            0         0                0        1          1         0   
20           0         1                0        1          0         0   
8            0         0                1        0          0         0   
14           0         0                1        1          1         0   

      count  
7   1890857  
15   683124  
54   140372  
5     93993  
23    88439  
4     85737  
12    80218  
13    60084  
31    52315  
1     47905  
62    23054  
6     12113  
20    10334  
8     10022  
14     9442

In [ ]:
#@title Co-ocorrência das variáveis com dados faltantes que podem ser imputadas
significant_missing_columns = ['occupation', 'employer']
co_occur = df[significant_missing_columns].isna().astype(int)
co_occur_agg = co_occur.groupby(significant_missing_columns).size().reset_index(name='count').sort_values(by='count', ascending=False)
co_occur_agg.head(15)

occupation  employer    count
0           0         0  2994621
3           1         1   179386
1           0         1   167864
2           1         0     6338

Dentre as variáveis com dados faltantes, nota-se que 'memo_cd', 'other_id' e 'memo_text' são majoritamente compostos por valores NaN. Entretanto, como são características não obrigatórias das contribuições, e suas co-ocorrências compõem uma grande parcela dos dados, assume-se que a ausência de registros seja dada apenas por um fator característico das doações.

Além disso, as instâncias que apresentam valores faltantes em variáveis com menos de 0,3% de dados ausentes ("state",
"city",
"entity_tp",
"transaction_dt",
"tran_id",
"name",
"file_num") podem ser descartadas.

A variável "transaction_pgi" possue 28.1% de dados faltantes e, para não haver interferências nas análises, optou-se por não realizar a imputação dessa variável. Assim, os valores NaN serão substituídos por "Desconhecido".

Por fim, as variáveis "employer" e "occupation" possuem porcentagens de dados faltantes que podem ser imputados.

Possíveis Abordagem: Como o propósito do presente trabalho é realizar análises interpretáveis dos dados, sem o enfoque em modelos preditivos e classificadores, deve-se tratar a imputação dos dados de tal forma que não haja perca de interpretabilida. Por isso, não se explora técnicas de imputação com modelos de machine learning convencionais uma vez que seria necessário a codificação das variáveis categóricas (Frequency encoding, Target encondig, One-Hot Enconding, Label Enconding e etc). Além disso, como grande parte das variáveis categóricas são de alta cardinalidade, uma codificação apropriada resultaria em uma base difícil de ser processada.

Por isso,



In [ ]:
  # Exclui-se as variáveis de indentificadores únicos
df.drop(columns=['name', 'sub_id', 'tran_id', 'image_num'], inplace=True)

# Exclui-se as variáveis 'memo_cd', 'other_id', 'memo_text'
df.drop(columns=['memo_cd', 'other_id', 'memo_text'], inplace=True)

In [ ]:
colunas_com_poucos_dados_faltantes = ["state", "city", "entity_tp", "transaction_dt", "file_num"]

df = df.dropna(subset=colunas_com_poucos_dados_faltantes)

In [ ]:
  df['transaction_pgi'] = df['transaction_pgi'].fillna("Desconhecido")

# Data Visualization

In [ ]:
colorscale = ['#0b132b', '#1b2740', '#2a3c56',
              '#354964', '#3a506b', '#3a506b',
              '#57b1b1', '#63d8cd', ' #6fffe9']

## 1. Perfis mais comums de Contribuintes

In [ ]:
columns = ['amndt_ind', 'rpt_tp', 'transaction_pgi', 'transaction_tp', 'entity_tp', 'occupation']
freq_table = df.groupby(columns).size().reset_index(name='counts')
freq_table = freq_table.sort_values('counts', ascending=False)

data = freq_table.head(9)
sankey = data.copy()

In [ ]:
sankey

amndt_ind rpt_tp transaction_pgi transaction_tp entity_tp occupation  \
250726         A     Q3               P             15       IND    RETIRED   
14653          A    12G               G             15       IND    RETIRED   
219901         A     Q2               P             15       IND    RETIRED   
71447          A    M10               G             15       IND    RETIRED   
182362         A     M9               P             15       IND    RETIRED   
45138          A    30G               G             15       IND    RETIRED   
233678         A     Q3               G             15       IND    RETIRED   
273042         A     YE               P             15       IND    RETIRED   
209278         A     Q2               P             15       IND   ATTORNEY   

        counts  
250726   29282  
14653    25330  
219901   24840  
71447    22537  
182362   20752  
45138    20589  
233678   17349  
273042   17077  
209278   14982

In [ ]:
#@title .

sankey = sankey[['transaction_tp', 'entity_tp', 'amndt_ind',
                 'occupation', 'transaction_pgi', 'rpt_tp', 'counts']]

original_attributes = ['transaction_tp', 'entity_tp', 'amndt_ind',
                       'occupation', 'transaction_pgi', 'rpt_tp']

display_attributes = ['Tipo de Transação', 'Tipo de Entidade', 'Indicador de Emenda',
                      'Ocupação', 'Etapa do Ciclo Eleitoral', 'Tipo de Relatório']

attribute_name_mapping = dict(zip(original_attributes, display_attributes))


node_labels = []
node_colors = []
node_attrs = []
node_indices = {}
color_palette = ["#eff7f6", "#a3abb2", "#707885", "#3e4658", "#242c42", "#0b132b",
                 "#2a3c56", "#354964", "#3a506b", "#57b1b1", "#63d8cd", "#6fffe9"]


attr_colors = {attr: color_palette[i % len(color_palette)] for i, attr in enumerate(original_attributes)}

current_index = 0
for attr in original_attributes:
    values = sankey[attr].unique()
    for val in values:
        node_name = str(val)
        if node_name not in node_indices:
            node_indices[node_name] = current_index
            node_labels.append(node_name)
            node_colors.append(attr_colors[attr])

            node_attrs.append(attribute_name_mapping[attr])
            current_index += 1


links = []
node_in_counts = {index: 0 for index in range(len(node_labels))}
node_out_counts = {index: 0 for index in range(len(node_labels))}

for _, row in sankey.iterrows():
    counts = row['counts']
    for i in range(len(original_attributes) - 1):
        source_attr = original_attributes[i]
        target_attr = original_attributes[i + 1]
        source_node = str(row[source_attr])
        target_node = str(row[target_attr])

        source_index = node_indices[source_node]
        target_index = node_indices[target_node]


        base_color = node_colors[source_index]
        link_color = base_color.replace('#', '0x')
        rgb_color = tuple(int(link_color[i:i+2], 16) for i in (2, 4, 6))
        rgba_color = f'rgba({rgb_color[0]}, {rgb_color[1]}, {rgb_color[2]}, 0.4)'

        links.append({
            'source': source_index,
            'target': target_index,
            'value': counts,
            'color': rgba_color
        })

        node_out_counts[source_index] += counts
        node_in_counts[target_index] += counts

node_counts = {}
for node_index in range(len(node_labels)):
    node_counts[node_index] = max(node_in_counts[node_index], node_out_counts[node_index])

sankey_nodes = {
    'label': node_labels,
    'color': node_colors,
    'customdata': list(zip(
        node_attrs,
        [node_counts[index] for index in range(len(node_labels))]
    )),
    'hovertemplate': '%{customdata[0]}: %{label}<br>Número de contribuintes: %{customdata[1]}<extra></extra>'
}

sankey_links = {
    'source': [link['source'] for link in links],
    'target': [link['target'] for link in links],
    'value': [link['value'] for link in links],
    'color': [link['color'] for link in links],
    'hovertemplate': (
        'Origem: %{source.label}<br>' +
        'Destino: %{target.label}<br>' +
        'Número de contribuintes: %{value}<extra></extra>'
    )
}

fig = go.Figure(go.Sankey(
    arrangement='snap',
    node=dict(
        pad=15,
        thickness=20,
        line=dict(color="black", width=0.5),
        label=sankey_nodes['label'],
        color=sankey_nodes['color'],
        customdata=sankey_nodes['customdata'],
        hovertemplate=sankey_nodes['hovertemplate']
    ),
    link=dict(
        source=sankey_links['source'],
        target=sankey_links['target'],
        value=sankey_links['value'],
        color=sankey_links['color'],
        hovertemplate=sankey_links['hovertemplate']
    )
))

# Título do gráfico
fig.update_layout(
    title_text="Perfis mais comuns dos contribuintes",
    font_size=12,
    paper_bgcolor="rgba(0,0,0,0)",
    font_color='black'
)

# Exibir o Sankey Diagram
fig.show()


In [ ]:
fig.write_html("sankey_diagram.html")

## 2. Proporção dos tipos de entidades, Indicadores de Emenda e tipos de transações.

In [ ]:
encoding = df['amndt_ind'].value_counts().reset_index().copy()
total = encoding['count'].sum()
encoding['percentage'] = encoding['count']/total
encoding

amndt_ind    count  percentage
0         A  2221016    0.665416
1         N  1114670    0.333955
2         T     2102    0.000630

In [ ]:
chart.save('setores.html')

In [ ]:
encoding = df['entity_tp'].value_counts().reset_index().copy()
total = encoding['count'].sum()
encoding['percentage'] = encoding['count']/total
encoding

entity_tp    count  percentage
0       IND  3313767    0.992803
1       ORG     8668    0.002597
2       CAN     7752    0.002322
3       PAC     4304    0.001289
4       CCM     2539    0.000761
5       COM      665    0.000199
6       PTY       93    0.000028

In [ ]:
encoding = df['transaction_tp'].value_counts().reset_index().copy()
total = encoding['count'].sum()
encoding['percentage'] = encoding['count']/total
encoding

transaction_tp    count  percentage
0              15  2993606    0.896883
1             24T   106449    0.031892
2             15E    93029    0.027871
3             22Y    73878    0.022134
4              10    33113    0.009921
5             24I    24529    0.007349
6             15C    10215    0.003060
7              11     2615    0.000783
8             20Y      183    0.000055
9              19      149    0.000045
10            21Y       22    0.000007

## 3. Quais etapas do Ciclo Eleitoral recebem mais contribuições

In [ ]:
# Data enconding for vis
quantidade_contribuicoes = df['transaction_pgi'].value_counts().reset_index()
quantidade_contribuicoes.columns = ['transaction_pgi', 'quantidade_contribuicoes']

quantia_total = df.groupby('transaction_pgi')['transaction_amt'].sum().reset_index()
quantia_total.columns = ['transaction_pgi', 'quantia_total']

vis1 = pd.merge(quantidade_contribuicoes, quantia_total, on='transaction_pgi')
mapeamento = {
    'P': 'Primária (Interna)',
    'G': 'Geral',
    'S': 'Especial',
    'E': 'Eleição',
    'C': 'Convenção',
    'O': 'Outras',
    'R': 'Segundo Turno',
    'Desconhecido': 'Desconhecido',
    '0': '0'
}

vis1['transaction_pgi'] = vis1['transaction_pgi'].map(mapeamento)

vis1 = vis1.sort_values(by='quantidade_contribuicoes', ascending=False)
vis1['razao_contribuicao'] = vis1['quantia_total']/vis1['quantidade_contribuicoes']

vis1

transaction_pgi  quantidade_contribuicoes  quantia_total  \
0  Primária (Interna)                   1567747   9.854442e+06   
1        Desconhecido                    938543   5.721768e+06   
2               Geral                    793031   4.871143e+06   
3              Outras                     12702   8.401678e+04   
4           Convenção                     11310   7.124079e+04   
5            Especial                      7776   4.907751e+04   
6       Segundo Turno                      5921   3.831923e+04   
7             Eleição                       523   3.519932e+03   
8                   0                       235   1.852021e+03   

   razao_contribuicao  
0            6.285735  
1            6.096437  
2            6.142437  
3            6.614453  
4            6.298920  
5            6.311408  
6            6.471750  
7            6.730271  
8            7.880940

In [ ]:
import altair as alt

# Criação da seleção interativa para o highlight ao passar o mouse
highlight = alt.selection_point(on='mouseover', fields=['transaction_pgi'], empty='none')

# Escala de cor personalizada
color_scale = alt.Scale(
    domain=vis1['transaction_pgi'].unique(),
    range=colorscale
)

# Gráfico de barras com highlight ao passar o mouse
bar = alt.Chart(vis1).mark_bar().encode(
    x=alt.X('transaction_pgi', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q', title='Quantia Total de Contribuições (em USD)'),
    color=alt.condition(
        highlight,
        alt.Color('transaction_pgi:N', scale=color_scale, title='Etapa'),
        alt.value('lightgray')  # Cor padrão para barras não destacadas
    ),
    tooltip=[
        alt.Tooltip('transaction_pgi:N', title='Etapa do Ciclo'),
        alt.Tooltip('quantia_total:Q', title='Quantia total de contribuição'),
        alt.Tooltip('razao_contribuicao:Q', title='Proporção Quantia/Contribuintes')
    ]

).add_params(
    highlight
).properties(
    width=600,
    height=300,
    title='Quantia Total de Contribuições por Etapa do Ciclo Eleitoral'
)

# Texto com valor, sem alterações
text = alt.Chart(vis1).mark_text(
    align='center',
    baseline='middle',
    dy=-10,
    color='black'
).encode(
    x=alt.X('transaction_pgi:N', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q'),
    text=alt.Text('razao_contribuicao:Q', format='.2f')
)

# Combinação do gráfico de barras e texto com configuração de estilo
vis = (bar + text).configure_title(
    fontSize=18,
).configure_axis(
    labelFontSize=12,
    titleFontSize=14,
).configure_legend(
    labelFontSize=12,
    titleFontSize=14,
).configure(
    background="rgba(0, 0, 0, 0)"
).configure_view(
    fill='white'
).configure_axis(
    grid=False,
)

vis.display()


alt.LayerChart(...)

In [ ]:
vis.save('vis1.html')

In [ ]:
#@title 3.2
color_scale = alt.Scale(
    domain=vis1['transaction_pgi'].unique()[3:],
    range=colorscale[3:]
)

bar = alt.Chart(vis1).transform_filter(
    alt.FieldOneOfPredicate(field='transaction_pgi', oneOf=['Outras', 'Convenção', 'Especial', 'Segundo Turno', 'Eleição', '0'])
).mark_bar().encode(
    x=alt.X('transaction_pgi', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q', title='Quantia Total de Contribuições (em USD)'),
    tooltip=[
        alt.Tooltip('transaction_pgi:N', title='Etapa do Ciclo'),
        alt.Tooltip('quantia_total:Q', title='Quantia total de contribuição'),
        alt.Tooltip('razao_contribuicao:Q', title='Proporção Quantia/Contribuintes')],
    color=alt.Color('transaction_pgi:N', scale=color_scale, title='Etapa'),
).properties(
    width=600,
    height=300,
    title='Quantia Total de Contribuições por Etapa do Ciclo Eleitoral (sem classes majoritárias)'
)

text = alt.Chart(vis1).transform_filter(
    alt.FieldOneOfPredicate(field='transaction_pgi', oneOf=['Outras', 'Convenção', 'Especial', 'Segundo Turno', 'Eleição', '0'])
).mark_text(
    align='center',
    baseline='middle',
    dy=-10,
    color='black'
).encode(
    x=alt.X('transaction_pgi:N', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q'),
    text=alt.Text('razao_contribuicao:Q', format='.2f')
)

vis = (bar + text).configure_title(
    fontSize=18
).configure_axis(
    labelFontSize=12,
    titleFontSize=14
).configure_legend(
    labelFontSize=12,
    titleFontSize=14
).configure(
    background="rgba(0, 0, 0, 0)"
).configure_view(
    fill='white'
).configure_axis(
    grid=False
)
vis.display()

alt.LayerChart(...)

In [ ]:
vis.save('vis2.html')

In [ ]:
import altair as alt

# Escala de cor personalizada para as categorias selecionadas
color_scale = alt.Scale(
    domain=vis1['transaction_pgi'].unique()[3:],
    range=colorscale[3:]
)

# Seleção interativa para o destaque ao passar o mouse
highlight = alt.selection_point(on='mouseover', fields=['transaction_pgi'], empty='none')

# Gráfico de barras com filtro e efeito de highlight ao passar o mouse
bar = alt.Chart(vis1).transform_filter(
    alt.FieldOneOfPredicate(field='transaction_pgi', oneOf=['Outras', 'Convenção', 'Especial', 'Segundo Turno', 'Eleição', '0'])
).mark_bar().encode(
    x=alt.X('transaction_pgi', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q', title='Quantia Total de Contribuições (em USD)'),
    tooltip=[
        alt.Tooltip('transaction_pgi:N', title='Etapa do Ciclo'),
        alt.Tooltip('quantia_total:Q', title='Quantia total de contribuição'),
        alt.Tooltip('razao_contribuicao:Q', title='Proporção Quantia/Contribuintes')
    ],
    color=alt.condition(
        highlight,
        alt.Color('transaction_pgi:N', scale=color_scale, title='Etapa'),
        alt.value('lightgray')  # Cor padrão para barras não destacadas
    )
).add_params(
    highlight
).properties(
    width=600,
    height=300,
    title='Quantia Total de Contribuições por Etapa do Ciclo Eleitoral (sem classes majoritárias)'
)

# Texto com valores numéricos
text = alt.Chart(vis1).transform_filter(
    alt.FieldOneOfPredicate(field='transaction_pgi', oneOf=['Outras', 'Convenção', 'Especial', 'Segundo Turno', 'Eleição', '0'])
).mark_text(
    align='center',
    baseline='middle',
    dy=-10,
    color='black'
).encode(
    x=alt.X('transaction_pgi:N', title='Etapa do Ciclo Eleitoral', sort='-y'),
    y=alt.Y('quantia_total:Q'),
    text=alt.Text('razao_contribuicao:Q', format='.2f')
)

# Combinação das camadas e configurações de estilo
vis = (bar + text).configure_title(
    fontSize=18
).configure_axis(
    labelFontSize=12,
    titleFontSize=14
).configure_legend(
    labelFontSize=12,
    titleFontSize=14
).configure(
    background="rgba(0, 0, 0, 0)"
).configure_view(
    fill='white'
).configure_axis(
    grid=False
)

vis.display()


alt.LayerChart(...)

In [ ]:
vis.save('vis3.html')

### 3.4 Existem diferenças entre as distribuições apresentadas?

In [ ]:
df_analysis = df[['transaction_pgi', 'transaction_amt']].copy()

groups = [group['transaction_amt'].values for name, group in df_analysis.groupby('transaction_pgi')]

stat, p = kruskal(*groups)
print(f'Estatística H de Kruskal-Wallis: {stat:.4f}')
print(f'p-valor: {p:.4e}')

Estatística H de Kruskal-Wallis: 38497.3239
p-valor: 0.0000e+00


In [ ]:
post_hoc = sp.posthoc_dunn(df_analysis, val_col='transaction_amt', group_col='transaction_pgi', p_adjust='fdr_bh')
pares_significantes = (post_hoc < 0.05)
significancia_dos_estados = pares_significantes.sum(axis=1)
significancia_dos_estados.sort_values(ascending=False)

0               8
Desconhecido    8
G               8
O               8
P               8
C               7
E               7
R               7
S               7
dtype: int64

Para comparar as distribuições acima, utiliza-se o teste estatístico **Kruskal-Wallis** com o intuito de verificar se existem diferenças significativas entre as medianas desses grupos. Sob a Hipótese nula (H0) de que as medianas dos grupos são iguais, obtem-se o p-valor < 0.05 e, assim, H0 é rejeitado.

Assim, para entender onde está a diferença entre esses grupos, faz-se o teste post-hoc de Dunn. Ele realiza comparações múltiplas entre pares dos grupos, aplicando testes Mann-Whitney U para cada par, ajustando os p-valores para múltiplas comparações.

Apesar dos resultados mostrarem que existem diferenças entre as distribuições das etapas do ciclo Eleitoral, procura-se entender o efeito prático dessa diferença: verificar se a diferença entre as distribuições é relevante nesse contexto. Para isso, calcula-se medidas de Tamanho de efeito.

In [ ]:
n_total = len(df_analysis)
k = len(groups)
epsilon_squared = (stat - k + 1) / (n_total - k)
print(f'Epsilon quadrado: {epsilon_squared:.6f}')

Epsilon quadrado: 0.011531


O Epsilon quadrado representa a proporção da variabilidade total dos dados que é explicada pela variável independente, nesse caso, as etapas de ciclo eleitoral. Seu intervalo varia de 0 a 1 e pelo seu cálculo, percebe-se que o tamanho do efeito das diferenças das distribuições é insignificante.

In [ ]:
def calculate_effect_size(group1, group2):
    n1 = len(group1)
    n2 = len(group2)
    u_statistic, _ = mannwhitneyu(group1, group2, alternative='two-sided')
    u_max = n1 * n2
    effect_size = 1 - (2 * u_statistic) / u_max
    return abs(effect_size)

group_dict = {
    name: group['transaction_amt'].values
    for name, group in df_analysis.groupby('transaction_pgi')
}

group_names = list(group_dict.keys())
n_groups = len(group_names)

effect_size_array = np.zeros((n_groups, n_groups))

for i in range(n_groups):
    group1 = group_names[i]
    data1 = group_dict[group1]

    for j in range(i, n_groups):
        group2 = group_names[j]
        data2 = group_dict[group2]
        effect_size = calculate_effect_size(data1, data2)
        effect_size_array[i, j] = effect_size
        effect_size_array[j, i] = effect_size

# Converter a matriz NumPy em DataFrame
effect_size_matrix = pd.DataFrame(
    effect_size_array,
    index=group_names,
    columns=group_names
)


In [ ]:
#@title .
effect_size_long = effect_size_matrix.reset_index().melt(
    id_vars='index',
    var_name='Grupo 2',
    value_name='Tamanho de Efeito'
)
effect_size_long.rename(columns={'index': 'Grupo 1'}, inplace=True)

# Criar o heatmap
heatmap = alt.Chart(effect_size_long).mark_rect().encode(
    x=alt.X('Grupo 2:N', title='Etapa do Ciclo Eleitoral '),
    y=alt.Y('Grupo 1:N', title='Etapa do Ciclo Eleitoral'),
    color=alt.Color('Tamanho de Efeito:Q', scale=alt.Scale(scheme='lightgreyteal')),
    tooltip=['Grupo 1', 'Grupo 2', alt.Tooltip('Tamanho de Efeito:Q', format='.4f')]
).properties(
    width=400,
    height=350,
    title='Matriz de Tamanho de Efeito entre Etapas do Ciclo Eleitoral'
)

text= alt.Chart(effect_size_long).mark_text(baseline='middle').encode(
    x='Grupo 2:O',
    y='Grupo 1:O',
    text=alt.Text('Tamanho de Efeito:Q', format='.2f'),
)

vis = (heatmap + text).configure_title(
    fontSize=18,

).configure_axis(
    labelFontSize=12,
    titleFontSize=14
).configure_legend(
    labelFontSize=12,
    titleFontSize=14
).configure(
    background="rgba(0, 0, 0, 0)"
).configure_view(
    fill='white'
).configure_axis(
    labelAngle=-45
)

# Exibir o heatmap
vis.display()

alt.LayerChart(...)

In [ ]:
#@title .
# Remover a coluna '0' e a linha '0' da matriz de tamanhos de efeito
effect_size_matrix_filtered = effect_size_matrix.drop(index='0', columns='0')

# Transformar a matriz filtrada em formato longo
effect_size_long_filtered = effect_size_matrix_filtered.reset_index().melt(
    id_vars='index',
    var_name='Grupo 2',
    value_name='Tamanho de Efeito'
)
effect_size_long_filtered.rename(columns={'index': 'Grupo 1'}, inplace=True)

import altair as alt

# Criar o heatmap com a matriz filtrada
heatmap = alt.Chart(effect_size_long_filtered).mark_rect().encode(
    x=alt.X('Grupo 2:N', title='Etapa do Ciclo Eleitoral'),
    y=alt.Y('Grupo 1:N', title='Etapa do Ciclo Eleitoral'),
    color=alt.Color('Tamanho de Efeito:Q', scale=alt.Scale(scheme='lightgreyteal')),
    tooltip=['Grupo 1', 'Grupo 2', alt.Tooltip('Tamanho de Efeito:Q', format='.4f')]
).properties(
    width=400,
    height=350,
    title='Matriz de Tamanho de Efeito entre Etapas do Ciclo Eleitoral'
)

# Adicionar os valores de tamanho de efeito como texto no heatmap
text = alt.Chart(effect_size_long_filtered).mark_text(baseline='middle').encode(
    x='Grupo 2:N',
    y='Grupo 1:N',
    text=alt.Text('Tamanho de Efeito:Q', format='.2f'),
)

# Combinar o heatmap e o texto
vis = (heatmap + text).configure_title(
    fontSize=18,
).configure_axis(
    labelFontSize=12,
    titleFontSize=14,
    labelAngle=-45
).configure_legend(
    labelFontSize=12,
    titleFontSize=14
).configure(
    background="rgba(0, 0, 0, 0)"
).configure_view(
    fill='white'

).configure_axis(
    labelAngle=-45
)

# Exibir o heatmap
vis.display()


alt.LayerChart(...)

# Further Explorations

Para a segunda entrega, espera-se explorar como as variáveis estão distribuídas no espaço e no tempo. Busca-se comprender a autocorrelação espacial e encontrar padrões na Série.

Além disso, explora-se técnicas de redução de dimensionalidade considerando as muitas variáveis categóricas. A partir das embeddings das categorias pode ser possível criar representações semânticas para as classes, possivelmente criando um espaço de dimensão pequena no qual haja agrupamentos dos contribuintes.



# References

Thomas, Tressy, and Enayat Rajabi. "A systematic review of machine learning-based missing value imputation techniques." Data Technologies and Applications 55.4 (2021): 558-585.

https://www.fec.gov/campaign-finance-data/
committee-type-code-descriptions/  
https://www.fec.gov/campaign-finance-data/transaction-type-code-descriptions/
